### PR TITLE
feat(launch): accept --nhosts as an alternative to --nproc for --auto-retry

### PR DESCRIPTION
**Pull request by saforem2**
## Problem

`--auto-retry` refused to run without `--nproc`:

```
--auto-retry requires --nproc (-n/--np) to be set explicitly.
```

But the rank count isn't what the loop needs. It immediately converts it:

```python
ranks_per_node = ngpu_per_host or ezpz.get_gpus_per_node() or 1
nhosts_active  = _ranks_to_hosts(ngpus, ranks_per_node)   # ranks -> hosts
```

...and never uses `ngpus` again — only `nhosts_active` reaches `_resolve_auto_retry_allocation`. So anyone sizing a job in **nodes** (the usual way you write a PBS `select=`) had to convert to ranks just to satisfy a validation check, then have the code convert it straight back.

## Change

Accept `--nhosts` (`-nh` / `--nnodes`) as a first-class alternative:

```bash
ezpz launch --auto-retry --nhosts 43 -- python3 train.py   # new
ezpz launch --auto-retry --np 512    -- python3 train.py   # unchanged
```

Nothing downstream needed touching — the plumbing was already there:

- `--nhosts` is already a CLI flag (`cli/flags.py:419`)
- already passed `main()` → `launch(nhosts=...)` (`launch.py:969`)
- `pbs._infer_topology` already handles `ngpus=None` when `nhosts` is set (Case A, `pbs.py:328`)

The restriction was simply stricter than the machinery required.

## Semantics

- **`--nhosts` is used verbatim.** It is already a host count; applying `_ranks_to_hosts` to it would be wrong.
- **When both are given, `--nhosts` wins** for the active/spare split. The unchanged downstream `_infer_topology` consistency checks still flag an inconsistent `ngpus`/`nhosts` pair, rather than this code silently picking one and hiding the conflict.
- **Both guards relaxed in step** — parse-time (CLI) and `launch()`-time (library) — so direct API callers get the same contract.
- `--nhosts <= 0` raises with a clear message.

## Tests

10 new/updated in `tests/test_launch_autoretry.py`:

- parse-time acceptance across all five aliases (`--nhosts`, `-nh`, `--nnodes`, `--nhost`, `--nnode`)
- both flags together
- neither flag → clear error (message updated)
- library-level: `nhosts` used verbatim, `ngpus` still converts, `nhosts` wins when both given, non-positive `nhosts` rejected

Verified to bite: **10 fail** against the old restriction, 115/115 pass with the change. Full suite 1395 passed (1 pre-existing `gqa` failure — the #199 `atol=1e-6` FLASH-vs-MATH issue, identical on clean `main`).

## Docs

`docs/cli/launch/index.md`: retitled the "Required: explicit `--nproc`" section, updated the mermaid decision flow, corrected the spare-pool formula to `total_pbs_nodes - active_hosts`, and added the equivalence example. `--auto-retry`'s `--help` text updated too.

## Label

`release:minor` — new accepted input, no behavior change for existing `--nproc` users.

## Summary by Sourcery

Allow auto-retry launch jobs to specify active size via either host count or process count and update messaging accordingly.

New Features:
- Support --nhosts (and its aliases) as a first-class alternative to --nproc for --auto-retry active size configuration.

Enhancements:
- Treat explicit host count as the authoritative source for active hosts when both nhosts and nproc are provided, with validation for non-positive values.
- Relax auto-retry preconditions so both CLI and launch() API accept either nhosts or ngpus for sizing while preserving downstream topology checks.

Documentation:
- Refresh auto-retry CLI documentation and help text to describe the nhosts/nproc equivalence, updated decision flow, and corrected spare-node calculation based on active hosts.

Tests:
- Add and update auto-retry tests to cover nhosts-based sizing, combined nhosts/nproc usage, validation of missing or invalid size flags, and host-count derivation in the launch auto-retry flow.

### CODE REVIEW

**Comment by saforem2** [comment]
## Validated on Sunspot: 9/9 scenarios pass ✅

Ran the full `test_launch_auto_retry.pbs` driver on a 4-node Sunspot allocation. Both new scenarios pass on real hardware:

```
[Scenario H] PASS  --nhosts 1 honored verbatim, rc=0 in 5s
[Scenario I] PASS  refused at parse time (rc=1), error names both flags
```

Full suite — `All scenarios passed. ✅` (job 12472776), with A–G confirming no regression to the `--nproc` path.

### Getting there took two other fixes, both split out

The first run was **3/9**. Rather than relax H/I's `rc=0` assertions, I traced each failure:

1. **#205** — `cleanup()` raised during teardown on a broken torch, turning already-successful runs into exit 1 (`Execution finished with 0` yet `$? == 1`). Fixing it took the driver **3/9 → 8/9**. Verified NOT a regression from this PR by A/B on a real allocation: branch `f9960ae` and main `b9a7f60` both exited 1 on the identical command.
2. **#203** — module eviction broke `import torch` on compute nodes; eliminated the `libsycl: undefined symbol: urDeviceWaitExp` error entirely (1 occurrence → 0).

### One fix belongs here (`30983a6`)

The ALCF proxy exports sat **inside** the `curl`-fallback branch, which never fires when the repo ships `utils.sh`. So every scenario ran without a proxy on nodes with no outbound internet. Scenario F (the only one running real training) was the only one exposed: `wandb.init()` blocked 60s on api.wandb.ai, no output for 300s, watchdog SIGTERM → `rc=124` after ~835s. A network failure masquerading as an auto-retry failure — the loop behaved correctly throughout. Hoisted the exports above the `UTILS` resolution; F now passes in 42s.

### Branch hygiene

Force-pushed to strip #203/#205 commits that my local stacked-test merges had pulled in. This branch is now exactly 3 commits over `main`, touching only launch/flags/tests/docs/driver.

**Merge order:** #205 → #203 → #204 (each unblocks the next one's validation).

File: src/ezpz/launch.py [hunk 1]
```diff
@@ -839,7 +852,21 @@ def launch(
         # one rank per GPU) when not. `or 1` covers the degenerate
         # case where get_gpus_per_node() returns 0 on a non-GPU node.
         ranks_per_node = ngpu_per_host or ezpz.get_gpus_per_node() or 1
-        nhosts_active = _ranks_to_hosts(ngpus, ranks_per_node)
+        if nhosts is not None:
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch.py:867]: Confirmed and fixed. `ranks_per_node` was resolved eagerly before the nhosts/ngpus branch, so the nhosts route paid for GPU detection it never uses — an explicit host count already IS the answer. On a launcher/login node without torch that call raises `ModuleNotFoundError`, breaking a well-specified `--auto-retry --nhosts N`.

Moved the resolution inside the ngpus branch, its only consumer. New test monkeypatches `get_gpus_per_node` to raise `ModuleNotFoundError` and asserts the nhosts route still reaches the allocation split with the right host count — verified it **fails** against the eager version.

File: src/ezpz/launch.py [hunk 2]
```diff
@@ -295,12 +295,20 @@ def parse_args(argv: Optional[Sequence[str]] = None):
...
-        if getattr(args, "nproc", -1) <= 0:
+        # What auto-retry actually needs is the ACTIVE HOST COUNT, so
+        # it can split the allocation into active + spare. --nproc
+        # gives that indirectly (ranks -> hosts); --nhosts gives it
+        # directly. Accept either rather than forcing users to convert
+        # a node count into a rank count just to satisfy this check.
+        if (
+            getattr(args, "nproc", -1) <= 0
+            and getattr(args, "nhosts", -1) <= 0
+        ):
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/launch.py]: Fixed in `db9d612`. Both flags default to `-1`, so "not provided" and "provided but invalid" are distinguishable, and they now get different errors:

```
$ ezpz launch --auto-retry --nhosts 0 -- ...
--auto-retry: --nhosts must be > 0, got 0. ...
```

Raised as `SystemExit` at parse time (clean, no traceback) rather than reaching the `ValueError` in `launch()`. Covers `--np 48 --nhosts 0` too, since the per-flag check runs before the either-or check. 4 parametrized cases added.
